### PR TITLE
feat(backend): federation: separated dht-hub features in new dht-node modul

### LINKED ISSUE #2358: 🐛 [Bug] Investigate Hyperswarm connection trouble in docker
## 🐛 Bugreport

Investigate Hyperswarm connection trouble in docker.

Apparently there is trouble when running the hyperswarm in docker - it makes the maria db to lose connection.

Investigate and figure out how to solve. Does this problem occur on all maschines?

Background: the hyperswarm dht PR is now part of the master - by setting `GRADIDO_HUB` as dht-topic in the config you should find stage1.

### LINKED ISSUE #2509: 🚀 [Feature] (federation) separate dht-node as new modul
## 🚀 Feature

- shift all dht-hub logic from backend- / federation-modul in a new dht-node modul 
- remove unnecessary parts and files

### PR DESCRIPTION
**Pull request by clauspeterhuebner**
### Issues

- fixes #2509
- fixes #2358

### Todo

- [X] shift all dht-hub logic from backend- / federation-modul in a new dht-node modul
- [x] remove unnecessary parts and files

### CODE REVIEW

File: dht-node/.env.dist [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+# SET LOG LEVEL AS NEEDED IN YOUR .ENV
+# POSSIBLE VALUES: all | trace | debug | info | warn | error | fatal
+# LOG_LEVEL=info
+
+# Federation
+# if you set the value of FEDERATION_DHT_TOPIC, the DHT hyperswarm will start to announce and listen
+# on an hash created from this topic
+FEDERATION_DHT_TOPIC=GRADIDO_HUB
+# FEDERATION_DHT_SEED=64ebcb0e3ad547848fef4197c6e2332f
```
Comment (clauspeterhuebner) [inline dht-node/.env.dist:26]: how is it ensured for deployment on production, that a dht-seed is set? bare-metal or command-line/system-parameter?

File: dht-node/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+## We Cannot do `$(date -u +'%Y-%m-%dT%H:%M:%SZ')` here so we use unix timestamp=0
+ENV BUILD_DATE="1970-01-01T00:00:00.00Z"
+## We cannot do $(npm run version).${BUILD_NUMBER} here so we default to 0.0.0.0
+ENV BUILD_VERSION="0.0.0.0"
+## We cannot do `$(git rev-parse --short HEAD)` here so we default to 0000000
+ENV BUILD_COMMIT="0000000"
+## SET NODE_ENV
+ENV NODE_ENV="production"
+## App relevant Envs
+ENV PORT="4000"
```
Comment (clauspeterhuebner) [inline dht-node/Dockerfile:18]: 4000 is the port of the backend module, right?
dht modul should listen on port 5000

File: docker-compose.yml [hunk 3]
```diff
@@ -111,6 +111,41 @@ services:
...
+    build:
+      # since we have to include the entities from ./database we cannot define the context as ./backend
+      # this might blow build image size to the moon ?!
+      context: ./
+      dockerfile: ./dht-node/Dockerfile
+      target: production
+    networks:
+      - internal-net
+    ports:
+      - 4000:4000
```
Comment (clauspeterhuebner) [inline docker-compose.yml:129]: thd dht modul should use 5000, because backend uses the port 4000

**Review by clauspeterhuebner — COMMENTED** [review]
Request changes or at least open questions. see comments

Can't set the radio-button to "Request Changes"...

File: docker-compose.yml (hunk 3 above)
Comment (Elweyn) [reply docker-compose.yml:129]:
```suggestion
      - 5000:5000
```

File: dht-node/Dockerfile (hunk 2 above)
Comment (Elweyn) [reply dht-node/Dockerfile:18]:
```suggestion
ENV PORT="5000"
```

**Comment by Mogge** [comment]
Can we please clarify, if the federation API lives in the same folder as the DHT Hyperswarm. In my opinion Docker is not that important at this point. It is even better to start the DHT locally on the console and not automatically with docker.

**Review by Mogge — APPROVED** [review]

File: dht-node/src/config/index.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+const constants = {
+  DB_VERSION: '0059-add_hide_amount_to_users',
+  LOG4JS_CONFIG: 'log4js-config.json',
+  // default log level on production should be info
+  LOG_LEVEL: process.env.LOG_LEVEL || 'info',
+  CONFIG_VERSION: {
+    DEFAULT: 'DEFAULT',
+    EXPECTED: 'v1.2023-01-01',
+    CURRENT: '',
+  },
```
Comment (Elweyn) [inline dht-node/src/config/index.ts:14]: Like we discused with @ulfgebhardt would be nice to refactor this into the database.

File: dht-node/src/dht_node/index.test.ts [hunk 5]
```diff
@@ -0,0 +1,798 @@
...
+const socketOnMock = jest.fn().mockImplementation((key: string, callback) => {
+  socketEventMocks[key] = callback
+})
+
+const socketWriteMock = jest.fn()
+
+const nodeConnectMock = jest.fn().mockImplementation(() => {
+  return {
+    on: socketOnMock,
+    once: socketOnMock,
```
Comment (Elweyn) [inline dht-node/src/dht_node/index.test.ts:72]: Same mock object, is this wanted?

File: dht-node/src/dht_node/index.ts [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+            logger.warn(
+              `received more than max allowed length of data buffer: ${data.length} against 1141 max allowed`,
+            )
+            return
+          }
+          logger.info(`data: ${data.toString('ascii')}`)
+          const recApiVersions: CommunityApi[] = JSON.parse(data.toString('ascii'))
+
+          // TODO better to introduce the validation by https://github.com/typestack/class-validato
+          if (recApiVersions && Array.isArray(recApiVersions) && recApiVersions.length < 5) {
```
Comment (Elweyn) [inline dht-node/src/dht_node/index.ts:62]: Why change the followed logic in this case? Please test the negative case and log and return.
```suggestion
          if (!recApiVersions && !Array.isArray(recApiVersions) && recApiVersions.length > 4) {
            logger.warn(
              `received totaly wrong or too much apiVersions-Definition JSON-String: ${JSON.stringify(
                recApiVersions,
              )}`,
            )
            return
          }
```

File: dht-node/src/dht_node/index.ts [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+                .execute()
+              logger.info(`federation community upserted successfully...`)
+            }
+          } else {
+            logger.warn(
+              `received totaly wrong or too much apiVersions-Definition JSON-String: ${JSON.stringify(
+                recApiVersions,
+              )}`,
+            )
+          }
```
Comment (Elweyn) [inline dht-node/src/dht_node/index.ts:112]: Moved to top
```suggestion
```

**Review by Elweyn — COMMENTED** [review]
Works on my local environment, looks great for me just some small suggestions.

**Review by Mogge — APPROVED** [review]

File: dht-node/src/dht_node/index.ts (hunk 6 above)
Comment (ulfgebhardt) [reply dht-node/src/dht_node/index.ts]: fixes. Btw you also have to change `&&` to `||`. ;)

**Review by Elweyn — APPROVED** [review]
Thanks to have added this, looks great for me!

File: dht-node/src/dht_node/index.ts (hunk 6 above)
Comment (Elweyn) [reply dht-node/src/dht_node/index.ts]: Good catch!

**Pull request merged**